### PR TITLE
stl to nrrd on upload + volume geometries

### PR DESCRIPTION
**Pull request by GoldenAnpu**
**volume_annotation_api.py**
 - `def upload_paths`
    - additionally as option receives List with mask directories in which stored NRRD files with geometries
    - if directory "interpolations' exists, checks this dir and convert old STL to NRRD with `stl_converter.to_nrrd`, updates annotaition and project meta with new items if it needs with `_update_on_transfer `
 - `def _update_on_transfer `  - creates new Class, Objec and Figure for converted STL, then removes old ones from annotation
 - `def _prepare_convertation_paths`  - Check dir and create paths for NRRD files if STL files need to be converted.
----
**volume_figure_api.py** 
 - `def download_sf_geometries` - new method that helps to download geometries to the specified paths
 - `def load_sf_geometry` - this method is an updated copy of `def load_figure_data`  but stored in the right place
----
**mask_3d.py**
- `def _bytes_from_file` - reads geometry from a file and return its UUID hex and bytes in tuple
- `def _bytes_from_file_batch` - the same but for a few files, returns mapped dictionary - hex: bytes
----
**volume_project.py**
- ` class VolumeDataset` - add new folder name `mask_dir = "mask"` for storing geometries for  Mask3D
-  ` def get_mask_dir` and `def get_mask_path` - to get path for files on download
-  `download_volume_project` 
    -  change dir for Mask3D storage to `mask_dir`
    -  add STL converter and project files updater if project has an old STL interpolations
-  `def upload_volume_project` now collects masks from `mask_dir`
----
**key_id_map.py**
 - `def from_dict` - to create key_id_map from dict
 - `def add_tag_to` , `def add_object_to`, `def add_figure_to`, `def add_video_to` - fixed, they used wrong method names
 - `def merge_key_id_maps` - method that helps to merge two maps with priority for duplicates
----
**nrrd_loader.py**
 - removed all Cyrillic characters
----
**stl_converter.py**
 - stores new functionality to properly convert old STL to NRRD: 
     -  `def matrix_from_nrrd_header`
     -  `def voxels_to_mask`
     -  `def to_nrrd`
----
**volume.py**
- fix typo 'nnrd'
----
**volume_figure.py** 
-  `def from_mask3d` - creates a VolumeFigure from Mask 3D geometry.
- `def _set_3d_geometry` - change the Mask3D geometry data in the figure.

### CODE REVIEW

File: supervisely/api/volume/volume_annotation_api.py [hunk 1]
```diff
@@ -109,7 +117,13 @@ def download(self, volume_id: int):
         volume_info = self._api.volume.get_info_by_id(volume_id)
         return self._download(volume_info.dataset_id, volume_id)
 
-    def append(self, volume_id: int, ann: VolumeAnnotation, key_id_map: KeyIdMap = None):
+    def append(
+        self,
+        volume_id: int,
+        ann: VolumeAnnotation,
+        key_id_map: KeyIdMap = None,
+        sf_geometries: Dict = None,
```
[the remark this answers is not in the document]
Comment (mkolomeychenko) [reply supervisely/api/volume/volume_annotation_api.py:125]: Dict Types

File: supervisely/api/volume/volume_annotation_api.py [hunk 2]
```diff
@@ -119,6 +133,8 @@ def append(self, volume_id: int, ann: VolumeAnnotation, key_id_map: KeyIdMap = N
         :type ann: VolumeAnnotation
         :param key_id_map: KeyIdMap object.
         :type key_id_map: KeyIdMap, optional
+        :param sf_geometries: Dict where keys are hex of sf.key(), and values are geometries, which represented as NRRD in byte format
+        :type key_id_map: Dict, optional
```
Comment (mkolomeychenko) [inline supervisely/api/volume/volume_annotation_api.py:137]: ?

File: supervisely/api/volume/volume_annotation_api.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mkolomeychenko) [reply supervisely/api/volume/volume_annotation_api.py:125]: ask Umar to change default class shape in DICOM volumes to Mask3D

use case 1 - main
obj = VolumeObject(tumor_class, mask_3d =  np3d_or_"~/home/mask.nrrd", tags=[])
ann = VolumeAnnotation(volume_size, objects=[obj1, obj2, ...], tags=[])
ann = ann.add_objects(objects=[a1, a2, a3])

api.volume.annotation.append(volume_id, ann)
api.volume.annotation.append_objects(volume_id, objects=[a1, a2, a3])

use case 2
obj = VolumeObject(tumor_class)
obj.add_mask_2d(plane_name: Literal["sag", "cor" ....], slice=123, sly.Bitmap() or "dsdasd.png" or np array 01-TF)


### obj.add_2d_label(plane_name: Literal["sag", "cor" ....], slice=123, sly.XXXGeometry)


ann = VolumeAnnotation(volume_size, objects=[obj1, obj2, ...])
api.volume.annotation.append(ann)
api.volume.annotation.append_objects(objects=[a1, a2, a3])

### later:
### VolumeFigure - actual only for 2d shapes
### obj.add_2d_figure(sly.Rectangle(dfsfdsf), plane, slice)


------------------------------------------------------
ann = VolumeAnnotation(...)
obj = VolumeObject(tumor_class)
tumor_mask_3d = sly.Mask3d(np 3d array //// "~/home/mask.nrrd" //// )
rectagle = sly.Rectangle(...)



obj.add_figure(tumor_mask_3d)

**Review by mkolomeychenko — CHANGES_REQUESTED** [review]

File: supervisely/io/fs.py [hunk 3]
```diff
@@ -1049,3 +1050,47 @@ def check_function(directory) -> bool:
...
+        from supervisely.io.fs import get_nested_dicts_data
+        nested_dict = {
+            'first_level': {
+                'second_level': {
+                    'third_level_1': 'You are wrong',
+                    'third_level_2': 'This is the value you are looking for'
+                }
+            }
+        }
+        required_value = get_nested_dicts_data(nested_dict, 'first_level', 'second_level', 'third_level_2')
```
Comment (mkolomeychenko) [inline supervisely/io/fs.py:1073]: nested_dict['first_level']['second_level']['third_level_2']

Comment (mkolomeychenko) [reply supervisely/io/fs.py:1073]: nested_dict.get('first_level', {}).get(['second_level', {}).get('third_level_2', {})

**Review by mkolomeychenko — CHANGES_REQUESTED** [review]

File: supervisely/api/volume/volume_annotation_api.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (GoldenAnpu) [reply supervisely/api/volume/volume_annotation_api.py]: добавил еще и api.project.append_classes(porject_id, [new_class])

File: supervisely/api/volume/volume_annotation_api.py (hunk 2 above)
Comment (GoldenAnpu) [reply supervisely/api/volume/volume_annotation_api.py]: убрал вовсе, теперь фигуры приходят с 3д аннотациями, геометрии забираются из них и аплодятся уже в самом конце глубоко внутри

File: supervisely/io/fs.py (hunk 3 above)
Comment (GoldenAnpu) [reply supervisely/io/fs.py]: removed at all

File: supervisely/api/volume/volume_annotation_api.py [hunk 4]
```diff
@@ -13,6 +16,11 @@
 
 from supervisely.api.entity_annotation.entity_annotation_api import EntityAnnotationAPI
 from supervisely.io.json import load_json_file
+from supervisely.io.fs import dir_exists, get_file_name
+from supervisely.volume import stl_converter
+import supervisely
```
[the remark this answers is not in the document]
Comment (GoldenAnpu) [reply supervisely/api/volume/volume_annotation_api.py]: fixed

File: supervisely/api/volume/volume_annotation_api.py [hunk 5]
```diff
@@ -193,20 +207,243 @@ def upload_paths(
...
+        :type key_id_map: KeyIdMap
+        :return: Updated ann, project_meta
+        :rtype: Tuple[VolumeAnnotation, ProjectMeta]
+        :Usage example:
+        """
+        # create unique ids for new objects
+        if nrrd_paths:
+            max_f_id = max(key_id_map.to_dict().get("figures").values())
+            max_o_id = max(key_id_map.to_dict().get("objects").values())
+            new_f_id = max_f_id + (10 ** (len(str(max_f_id)) - 1)) + random.randint(100, 999)
```
Comment (mkolomeychenko) [inline supervisely/api/volume/volume_annotation_api.py:352]: server ids can not be generated on client side?

File: supervisely/api/volume/volume_figure_api.py [hunk 6]
```diff
@@ -142,23 +147,25 @@ def append_bulk(self, volume_id: int, figures: List[VolumeFigure], key_id_map: K
             api.volume.figure.append_bulk(volume_id, [figure], key_id_map)
         """
 
-        if len(figures) == 0:
+        if not figures:
```
Comment (mkolomeychenko) [inline supervisely/api/volume/volume_figure_api.py:150]: ?

File: supervisely/api/volume/volume_figure_api.py [hunk 7]
```diff
@@ -278,57 +285,6 @@ def interpolate(self, volume_id: int, spatial_figure: VolumeFigure, key_id_map:
         )
         return response.content
 
-    # def interpolate_batch(
```
Comment (mkolomeychenko) [inline supervisely/api/volume/volume_figure_api.py:281]: keep

File: supervisely/api/volume/volume_figure_api.py [hunk 8]
```diff
@@ -434,22 +390,22 @@ def _append_bulk_mask3d(
         :rtype: :class:`NoneType`
         :Usage example:
         """
-        figures_count = len(figures_json)
+        figures_count = len(figures)
         if figures_count == 0:
```
Comment (mkolomeychenko) [inline supervisely/api/volume/volume_figure_api.py:394]: if len() == 0

File: supervisely/api/volume/volume_figure_api.py [hunk 9]
```diff
@@ -459,25 +415,172 @@ def _append_bulk_mask3d(
...
+            ann = sly.VolumeAnnotation.from_json(vol_ann_json, project_meta, key_id_map)
+            spatial_figures = [sp_figure.key() for sp_figure in ann.spatial_figures]
+            for figure in spatial_figures:
+                geometries[figure] = geometry_bytes
+            api.volume.figure.upload_sf_geometries(spatial_figures, geometries, key_id_map)
         """
 
-        for sf, geometry_bytes in zip(spatial_figures, geometries):
-            figure_id = key_id_map.get_figure_id(sf.key())
+        if not spatial_figures:
```
Comment (mkolomeychenko) [inline supervisely/api/volume/volume_figure_api.py:467]: ?

File: supervisely/geometry/mask_3d.py [hunk 10]
```diff
@@ -255,27 +255,75 @@ def geometry_name():
...
-        figure.geometry._space = mask3d_header["space"]
-        figure.geometry._space_directions = mask3d_header["space directions"]
-        path_without_filename = "/".join(file_path.split("/")[:-1])
-        remove_dir(path_without_filename)
+        geometry._space = mask3d_header["space"]
+        geometry._space_directions = mask3d_header["space directions"]
+        return geometry
+
+    @staticmethod
+    def to_figure_from_file(figure, file_path: str):
```
Comment (mkolomeychenko) [inline supervisely/geometry/mask_3d.py:278]: from_file

File: supervisely/geometry/mask_3d.py [hunk 11]
```diff
@@ -255,27 +255,75 @@ def geometry_name():
...
+        :type figure: VolumeFigure
+        :param file_path: Path to nrrd file with data
+        :type file_path: str
+        """
+        if not isinstance(figure.geometry, Mask3D):
+            raise Exception(f"You trying to add Mask3D geometry to {figure.geometry.name()}")
+        figure.geometry.data = Mask3D.from_file(file_path).data
+
+    @staticmethod
+    def to_figure_from_bytes(figure, geometry_bytes: bytes):
```
Comment (mkolomeychenko) [inline supervisely/geometry/mask_3d.py:292]: from_bytes

File: supervisely/geometry/mask_3d.py [hunk 12]
```diff
@@ -255,27 +255,75 @@ def geometry_name():
...
+            raise Exception(f"You trying to add Mask3D geometry to {figure.geometry.name()}")
+
+        with tempfile.NamedTemporaryFile(delete=True) as temp_file:
+            temp_file.write(geometry_bytes)
+            data_array, _ = nrrd.read(temp_file.name)
+
+        figure.geometry.data = data_array
+
+    @classmethod
+    def to_figure_from_array(cls, figure, data_array: np.ndarray) -> None:
```
Comment (mkolomeychenko) [inline supervisely/geometry/mask_3d.py:312]: from_np

**Review by mkolomeychenko — CHANGES_REQUESTED** [review]

File: supervisely/volume/stl_converter.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+    ] = mask[
+        vol_inter_max[0] - start[0] : vol_inter_min[0] - start[0],
+        vol_inter_max[1] - start[1] : vol_inter_min[1] - start[1],
+        vol_inter_max[2] - start[2] : vol_inter_min[2] - start[2],
+    ]
+    padded_mask = padded_mask.astype(dtype="uint8")
+    return padded_mask
+
+
+def save_to_nrrd_file_on_upload(interpolation_dir: str) -> List[str]:
```
Comment (mkolomeychenko) [inline supervisely/volume/stl_converter.py:126]: convert_str_to_nrrd

Comment (mkolomeychenko) [reply supervisely/volume/stl_converter.py:126]: str_to_nrrd

File: supervisely/volume/stl_converter.py [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+    logger.warning(
+        f"STL format is no longer supported. STL is automatically converted to NRRD containing 3D Mask and moved to 'old_interpolation' directory"
+    )
+
+    nrrd_paths = []
+
+    for file in mesh_paths:
+        stl_path = os.path.join(file)
+        nrrd_path = stl_path.replace(_stl_ext, _nrrd_ext)
+        volume_path = os.path.dirname(stl_path.replace(interpolation_dir_name, volume_dir_name))
```
Comment (mkolomeychenko) [inline supervisely/volume/stl_converter.py:193]: interpolation_dir_name - ???

File: supervisely/volume_annotation/volume_object.py [hunk 15]
```diff
@@ -38,4 +50,39 @@ class VolumeObject(VideoObject):
...
+            obj_class=obj_class,
+            tags=tags,
+            key=key,
+            class_id=class_id,
+            labeler_login=labeler_login,
+            updated_at=updated_at,
+            created_at=created_at,
+        )
+
+        if mask_3d:
```
Comment (mkolomeychenko) [inline supervisely/volume_annotation/volume_object.py:74]: is not None

File: supervisely/volume_annotation/volume_annotation.py [hunk 16]
```diff
@@ -658,3 +659,50 @@ def to_json(self, key_id_map: KeyIdMap = None) -> dict:
...
+        if isinstance(objects, List):
+            objects = VolumeObjectCollection(objects)
+
+        # check if objects without figures
+        for _, vobject in objects._collection.items():
+            try:
+                vobject.figure
+            except AttributeError as e:
+                e.args = [
+                    "There is no spatial figure in 'VolumeObject'",
```
Comment (mkolomeychenko) [inline supervisely/volume_annotation/volume_annotation.py:700]: 3D mask for object is not defined

File: supervisely/volume_annotation/volume_annotation.py [hunk 17]
```diff
@@ -658,3 +659,50 @@ def to_json(self, key_id_map: KeyIdMap = None) -> dict:
...
+                ]
+                raise e
+
+        sf_figures = [vobject.figure for vobject in objects]
+        collection = self.objects.add_items(objects)
+        self = self.clone(objects=collection)
```
Comment (mkolomeychenko) [inline supervisely/volume_annotation/volume_annotation.py:706]: new_ann = self.clone(objects=collection)
new_ann.spatial_figures.extend(sf_figures)
return new_ann

File: supervisely/volume/stl_converter.py [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+import trimesh
+import numpy as np
+
+from typing import List, Dict, Any
+from supervisely import logger
+from supervisely.io.fs import get_file_name_with_ext, file_exists, dir_exists, mkdir
+
+_stl_ext = ".stl"
+_nrrd_ext = ".nrrd"
+volume_dir_name = "volume"
```
Comment (mkolomeychenko) [inline supervisely/volume/stl_converter.py:16]: hardcode

File: supervisely/volume/stl_converter.py (hunk 13 above)
Comment (mkolomeychenko) [reply supervisely/volume/stl_converter.py:126]: stl_to_nrrd(stl_path, nrrd_path):
....

File: supervisely/api/volume/volume_annotation_api.py [hunk 19]
```diff
@@ -161,6 +172,7 @@ def upload_paths(
         project_meta: ProjectMeta,
         interpolation_dirs=None,
         progress_cb: Optional[Union[tqdm, Callable]] = None,
+        mask_dirs=None,
```
Comment (mkolomeychenko) [inline supervisely/api/volume/volume_annotation_api.py:175]: remove?

**Review by mkolomeychenko — CHANGES_REQUESTED** [review]

File: supervisely/volume_annotation/volume_figure.py [hunk 20]
```diff
@@ -582,3 +603,68 @@ def to_json(self, key_id_map=None, save_meta=True):
...
+            volume_object,
+            mask_3d,
+            key=key,
+            class_id=class_id,
+            labeler_login=labeler_login,
+            updated_at=updated_at,
+            created_at=created_at,
+        )
+
+    def set_geometry(self, new_geometry: Mask3D) -> None:
```
Comment (mkolomeychenko) [inline supervisely/volume_annotation/volume_figure.py:654]: _set_3d_geometry

File: supervisely/volume_annotation/volume_figure.py [hunk 21]
```diff
@@ -582,3 +603,68 @@ def to_json(self, key_id_map=None, save_meta=True):
...
+        """
+        Change the Mask3D geometry data in the figure.
+
+        :param new_geometry: The new Mask3D geometry object.
+        :type new_geometry: Mask3D
+        """
+        if new_geometry.name() != Mask3D.name():
```
Comment (mkolomeychenko) [inline supervisely/volume_annotation/volume_figure.py:661]: type

Comment (mkolomeychenko) [reply supervisely/volume_annotation/volume_figure.py:661]: isinstance

File: supervisely/project/volume_project.py [hunk 22]
```diff
@@ -360,42 +390,13 @@ def download_volume_project(
...
-                    mesh_paths.append(figure_path)
-                api.volume.figure.download_stl_meshes(mesh_ids, mesh_paths)
             if log_progress:
                 ds_progress.iters_done_report(len(batch))
             if progress_cb is not None:
                 progress_cb(len(batch))
     project_fs.set_key_id_map(key_id_map)
 
 
-def load_figure_data(
```
Comment (mkolomeychenko) [inline supervisely/project/volume_project.py:378]: why?

File: supervisely/geometry/mask_3d.py [hunk 23]
```diff
@@ -354,7 +367,7 @@ def from_json(cls, json_data: Dict) -> Mask3D:
             figure = sly.Mask3D.from_json(figure_json)
         """
         if json_data == {}:
-            return cls(data=np.zeros((3, 3, 3), dtype=np.bool_))
+            return cls(data=np.random.randint(2, size=(3, 3, 3), dtype=np.bool_))
```
Comment (mkolomeychenko) [inline supervisely/geometry/mask_3d.py:370]: ?

**Review by mkolomeychenko — CHANGES_REQUESTED** [review]

**Comment by GoldenAnpu** [comment]
Create new PR based on this one after https://github.com/supervisely/supervisely/pull/699 will be merged.
Need to transfer just changes that described in this PR

**Comment by GoldenAnpu** [comment]
Merge this after https://github.com/supervisely/supervisely/pull/699 will be merged.

**Pull request closed without merging**